### PR TITLE
Covergrid Browser

### PR DESCRIPTION
**Pull request by darthoctopus**
Updated @brunob's covergrid to work with latest commit (see #241)

### CODE REVIEW

**Comment by lazka** [comment]
Thanks. As you can see tests are failing, see https://quodlibet.readthedocs.io/en/latest/development/testing.html for more details.

For the browser in general, I think it's still a bit rough, and performance is horrible, but maybe including it will push us to fix things and prevent further bitrot

@declension any thoughts?

**Comment by darthoctopus** [comment]
could you elaborate on what you mean by being rough? e.g. UI not sufficiently polished? I think one sorely needed thing (not sure how to implement this) would be the ability to set a different cover size from the one used by the Album browser in GUI preferences (perhaps also toggle between RHPaned and RVPaned, not sure how to implement either)

File: quodlibet/quodlibet/browsers/covergrid/prefs.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+from quodlibet import _
+from quodlibet.browsers._base import FakeDisplayItem, EditDisplayPatternMixin
+from quodlibet.formats import PEOPLE
+from quodlibet.qltk import Button, Icons
+from quodlibet.qltk.ccb import ConfigCheckButton
+from quodlibet.util import format_rating
+from quodlibet.util.i18n import numeric_phrase
+
+PEOPLE
+_SOME_PEOPLE = "\n".join([util.tag("artist"), util.tag("performer"),
```
Comment (declension) [inline quodlibet/quodlibet/browsers/covergrid/prefs.py]: Hmm we should really extract all this code, but that can be done separately

File: quodlibet/quodlibet/browsers/covergrid/prefs.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                         util.tag("composer"), util.tag("arranger"), ])
+
+_EMPTY = _("Songs not in an album")
+DEFAULT_PATTERN_TEXT = """[b]<album|<album>|%s>[/b]<date| (<date>)>
+[small]<~discs|<~discs> - ><~tracks> - <~long-length>[/small]
+<~people>""" % _EMPTY
+
+
+class Preferences(qltk.UniqueWindow, EditDisplayPatternMixin):
```
Comment (declension) [inline quodlibet/quodlibet/browsers/covergrid/prefs.py]: Does this in fact extend `browsers.albums.Prefs`? It looks almost identical

**Comment by declension** [comment]
@darthoctopus 

Firstly thanks for your (and @brunob's) work so far. It's definitely one of the most common requests for browsers and it's starting to look good.

### Fixes
* As to the current state, now that the tests are green, the most pressing problem I had was on clicking the Preferences menu item:

```
Traceback (most recent call last):
  File "/home/nick/workspace/quodlibet/quodlibet/quodlibet/util/__init__.py", line 753, in wrap
    return handler(that, *args)
  File "/home/nick/workspace/quodlibet/quodlibet/quodlibet/browsers/covergrid/prefs.py", line 62, in __init__
    cb.set_active(config.getboolean("browsers", "album_text"))
  File "/home/nick/workspace/quodlibet/quodlibet/quodlibet/util/config.py", line 173, in getboolean
    raise Error(e)
ConfigParser.Error: <exception str() failed>
```
* The display pattern doesn't seem to be centred properly relative to the Album Art (when large): 
covergrid alignment issue


### Performance
I agree with @lazka really - the performance is still the single killer for this PR. On this machine (4GHz I7) it takes:

 *  7-8 seconds to clear a query to none (i.e. redraw all albums)
 *  7-8 seconds to re-render the window on resize with no query selected

Both are occasionally enough for the Unity to darken the window, i.e. indicate an unresponsive UI.

I've noticed that as soon as the query is specific enough (e.g. `artist=nas`) to narrow down the result set a lot, the performance is actually very fast. You can see this happen in realtime as the first letters are  typed (`artist=n`). (at the very least, this browser should override the timeout for query updating).

I have around 8500 songs (across 1800 albums) in the library, which is an order of magnitude less than some of our users.

### Discussion
* _Should_ this be a subclass of AlbumList (currently seems like it works well as such)
* Perhaps then, the prefs should subclass its prefs too as there's large overlap.
* Maybe we could pad the browser window, not sure.
* Does the _All Albums_ item make sense in this browser?

Thanks

**Comment by lazka** [comment]
> ### Discussion
> * _Should_ this be a subclass of AlbumList (currently seems like it works well as such)

I'd avoid subclassing browsers, this has always made things complicated in the past. 

> * Perhaps then, the prefs should subclass its prefs too as there's large overlap.

Same, at least until we are happy with how things are.

but ymmv

Regarding the performance, the only way I know to make things faster with iconview (which has same nasty O(n^3)) is to move the text inside the cover and make the cover size fixed. Another might be to use Gtk.FlowBox (not available on Ubuntu 14.04)

**Comment by declension** [comment]
> > Should this be a subclass of AlbumList (currently seems like it works well as such)

> I'd avoid subclassing browsers, this has always made things complicated in the past.

I tend to agree actually. @darthoctopus probably better to start a _new_ browser here, extract any _necessarily_ common code to mixins.

File: quodlibet/quodlibet/browsers/covergrid/prefs.py (hunk 2 above)
Comment (darthoctopus) [inline quodlibet/quodlibet/browsers/covergrid/prefs.py]: I'm not sure how to properly extend browsers.albums.Preferences without also invoking the AlbumList preferences.

**Comment by darthoctopus** [comment]
> this browser should override the timeout for query updating

where can I find documentation on this? Sorry, I'm quite new to both Gtk and quodlibet

**Comment by darthoctopus** [comment]
I've implemented @lazka's suggestion to fix item width, could you try again to see if this improves performance? I've also fixed the centering issue.

**Comment by lazka** [comment]
Still not really usable here, but if all fails we can disable it for the next release.

**Comment by declension** [comment]
Thanks all

**Comment by brunob** [comment]
(y)

**Comment by urielz** [comment]
Just chiming in my feedback. I have a relatively small library (around 250 albums) and it is very slow on my end as well. However, I’d rather have this than no cover grid at all! So very happy to see this pr merged.
 
Question: are images cached? does the file size of images have an impact on performance? (All cover images I have are fairly big, .5 to 1Mb each, and nothing is embedded as metadata...)

**Comment by lazka** [comment]
> Question: are images cached? does the file size of images have an impact on performance? (All cover images I have are fairly big, .5 to 1Mb each, and nothing is embedded as metadata...)

thumbnails of external images are cached

**Pull request merged**